dvc/remote/base.py

```diff
@@ -312,7 +312,7 @@ def get_checksum(self, path_info):
             return checksum
 
         if self.isdir(path_info):
-            checksum = self.get_dir_checksum(path_info)
+            checksum = self.get_dir_checksum(path_info / "")
```

Comment:
This does not leak any GS internals. This could have been solved by adding it to methods inside `RemoteGS`, but, I made a compromise between handling it higher vs looking into each remotes and fixing there.

> Why do we need this?
Say, you have `gs://dvc-test/data1.txt` file and `gs://dvc-test/data/data2.txt`. If I do `dvc add gs://dvc-test/data`, it will try to `walk_file` which will include `gs://dvc-test/data1.txt` as well.

> Another questions is why don't we need this in S3? Is there a bug?
Not sure, if this is an issue in real, quick addition of `s3.put_object(Bucket="bucket", Key="data1.txt", Body=b"3")` in tests did fail the `test_walk_files` for s3. Needs further investigation. Probably a bug.

Reply:
@skshetry I agree with @Suor , this change looks very weird. `get_dir_checksum` should simply accept `path_info`, you area leaking "empty_dir" concept here. Effectively you are feeding it an empty_dir object and not the directory as a whole, which might be confusing and error prone. s3 didn't have to do this, so I'm pretty sure you don't have to do this as well.

Reply:
> Say, you have gs://dvc-test/data1.txt file and gs://dvc-test/data/data2.txt. If I do dvc add gs://dvc-test/data, it will try to walk_file which will include gs://dvc-test/data1.txt as well.

@skshetry, yep, this is due to S3 and GCS searching by prefix instead of paths, I think it is remote's problem.